plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/weighers/Weighers.kt

```diff
@@ -5,15 +5,98 @@ package org.jetbrains.kotlin.idea.completion.weighers
 import com.intellij.codeInsight.completion.CompletionSorter
 import com.intellij.codeInsight.lookup.LookupElement
 import org.jetbrains.kotlin.analysis.api.KtAnalysisSession
+import org.jetbrains.kotlin.analysis.api.ValidityTokenOwner
+import org.jetbrains.kotlin.analysis.api.components.KtImplicitReceiver
 import org.jetbrains.kotlin.analysis.api.symbols.KtSymbol
+import org.jetbrains.kotlin.analysis.api.tokens.ValidityToken
+import org.jetbrains.kotlin.analysis.api.types.KtSubstitutor
 import org.jetbrains.kotlin.analysis.api.types.KtType
+import org.jetbrains.kotlin.analysis.api.withValidityAssertion
+import org.jetbrains.kotlin.idea.completion.ImportableFqNameClassifier
+import org.jetbrains.kotlin.idea.project.TargetPlatformDetector
+import org.jetbrains.kotlin.idea.project.findAnalyzerServices
+import org.jetbrains.kotlin.idea.project.languageVersionSettings
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtExpression
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.ImportPath
+
+internal class WeighingContext private constructor(
+    override val token: ValidityToken,
+    val explicitReceiver: KtExpression?,
+    private val myExpectedType: KtType?,
+    private val myImplicitReceivers: List<KtImplicitReceiver>,
+    val importableFqNameClassifier: ImportableFqNameClassifier,
+) : ValidityTokenOwner {
+    val expectedType: KtType?
+        get() = withValidityAssertion {
+            myExpectedType
+        }
+
+    /** All implicit receivers in the current resolution context. The receiver declared in the inner most scope appears first. */
+    val implicitReceiver: List<KtImplicitReceiver>
+        get() = withValidityAssertion {
+            myImplicitReceivers
+        }
+
+    fun withoutExpectedType(): WeighingContext = withValidityAssertion {
+        WeighingContext(token, explicitReceiver, null, myImplicitReceivers, importableFqNameClassifier)
+    }
+
+    companion object {
+        fun KtAnalysisSession.createWeighingContext(
+            receiver: KtExpression?,
+            expectedType: KtType?,
+            implicitReceivers: List<KtImplicitReceiver>,
+            fakeCompletionFile: KtFile,
+        ): WeighingContext {
+            val defaultImportPaths = fakeCompletionFile.getDefaultImportPaths()
+            return WeighingContext(
+                token,
+                receiver,
+                expectedType,
+                implicitReceivers,
+                ImportableFqNameClassifier(fakeCompletionFile) { defaultImportPaths.hasImport(it) })
+        }
+
+        fun KtAnalysisSession.createEmptyWeighingContext(
+            fakeCompletionFile: KtFile
+        ): WeighingContext {
+            val defaultImportPaths = fakeCompletionFile.getDefaultImportPaths()
```

Comment:
MINOR: You can use `createWeighingContext` function here to minimize the duplication